lua_modules/http/httpserver.lua

```diff
@@ -23,54 +23,49 @@ do
   -- response methods
   ------------------------------------------------------------------------------
   local make_res = function(csend, cfini)
-   local send = function(self, data, status)
-    -- TODO: req.send should take care of response headers!
-    if self.send_header then
-      csend("HTTP/1.1 ")
-      csend(tostring(status or 200))
-      -- TODO: real HTTP status code/name table
-      csend(" OK\r\n")
-      -- we use chunked transfer encoding, to not deal with Content-Length:
-      --   response header
-      self:send_header("Transfer-Encoding", "chunked")
-      -- TODO: send standard response headers, such as Server:, Date:
+   -- send("hello world\r\n") will send the data
+    -- send() will close connection
+    local send = function(data)
+      if data then
+        -- chunked transfer encoding
+        csend(("%X\r\n"):format(#data))
+        csend(data)
+        csend("\r\n")
+      else
+        -- finalize chunked transfer encoding
+        csend("0\r\n\r\n")
+        -- close connection
+        cfini()
+      end
     end
-    if data then
-      -- NB: no headers allowed after response body started
-      if self.send_header then
-        self.send_header = nil
-        -- end response headers
+    local send_header = function(header)
```

Comment:
I think a better API would be for `send_header` to take the `status` and `headers` separately, rather than packaged together into a table.

Reply:
Is it 3 function?
one is the `send` just send the response body data.
other is `send_status` just send the first line `HTTP/1.1 200 OK\r\n`
and `send_header` just send the headers like `Connection: close`

Reply:
The old API, AFAICT, indeed had `status` passed to a different function, didn't it?

Maybe the three-function API is a better plan; would you mind seeing what it makes the example look like?

Reply:
the `send` is current version

```lua
-- status maybe `404` or `404 Not Found`
function send_status(status)
    status = status or 200
    csend(("HTTP/1.1 %s\r\n"):format(status_code[status] or status))
    csend("Transfer-Encoding: chunked\r\n")
end

function send_headers(headers)
    -- header
    for k, v in pairs(header.headers) do
      csend(k)
      csend(": ")
      csend(v)
      csend("\r\n")
    end
    csend("\r\n")
end
```

because `csend("\r\n")` in `send_headers`. so, a necessary operation is 
```lua
send_status()
send_headers()
send("hello world")
send()
```

Reply:
I think I would prefer a two-function API, but with a twist:

```lua
sendbody = res:send_headers(status, {foo = bar, ...})
sendbody(data)
-- ...
sendbody(nil)
```

That is, I would like `send_headers` to return the body-sending function; this way, there's no (or at least, less) ambiguity about which one the user is to call when, I think?

Reply:
great

Reply:
But what if the user does not need to send any headers?

Reply:
I think `res:send_headers()` send the 200 without any headers.

Reply:
I would prefer to have 3 methods right from the start.
`send_status` and `send_header`  (or combined in one function)
`send(body)`

I would prefer to be able to omit any of the calls so to just send an 200 OK response calling `send(nil)` would be sufficient.

The server would need to have a status to know when it has to internally send status and headers and also to prevent sending status or headers multiple times or in a wrong order.

Reply:
@HHHartmann: How often does one want to send an empty 200 response, and is `res:send_headers()()` really that bad if that's what you want?

Reply:
there has `\r\n` between response header and response data. if just send an 200. the `\r\n` need send alone.

so, I think that send_status and send_header need combined

Reply:
Still SendHeader needs to be called with a complex structure, Why do we change the interface in such a way without good reason? It will break everybodies code. Users will even have to adapt their logic if unlucky and collect all header data themselves to pass them at once. Aside from this multiple headers with the same key are not possible that way.